agent/connect/ca/provider_vault_auth_approle.go

```diff
@@ -0,0 +1,60 @@
+package ca
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/hashicorp/consul/agent/structs"
+)
+
+// left out 2 config options as we are re-using vault agent's auth config.
+// Why?
+// remove_secret_id_file_after_reading - don't remove what we don't own
+// secret_id_response_wrapping_path - wrapping the secret before writing to disk
+//                                    (which we don't need to do)
+
+func NewAppRoleAuthClient(authMethod *structs.VaultAuthMethod) (*VaultAuthClient, error) {
+	authClient := NewVaultAPIAuthClient(authMethod, "")
+	// check for hardcoded /login params
+	if legacyCheck(authMethod.Params, "role_id", "secret_id") {
+		return authClient, nil
+	}
+
+	// check for required config params
+	key := "role_id_file_path"
+	if val, ok := authMethod.Params[key].(string); !ok {
+		return nil, fmt.Errorf("missing '%s' value", key)
+	} else if strings.TrimSpace(val) == "" {
+		return nil, fmt.Errorf("'%s' value is empty", key)
+	}
+	authClient.LoginDataGen = ArLoginDataGen
+
+	return authClient, nil
+}
+
+// don't need to check for legacy params as this func isn't used in that case
+func ArLoginDataGen(authMethod *structs.VaultAuthMethod) (map[string]any, error) {
+	params := authMethod.Params
+	// role_id is required
+	roleIdFilePath := params["role_id_file_path"].(string)
+	// secret_id is optional (secret_ok is used in check below)
+	secretIdFilePath, secret_ok := params["secret_id_file_path"].(string)
+
+	var err error
+	var rawRoleID, rawSecretID []byte
+	data := make(map[string]any)
+	if rawRoleID, err = os.ReadFile(roleIdFilePath); err != nil {
+		return nil, err
+	}
+	data["role_id"] = string(rawRoleID)
+	switch rawSecretID, err = os.ReadFile(secretIdFilePath); {
+	case err != nil && secret_ok:
+		return nil, err
+	case len(bytes.TrimSpace(rawSecretID)) > 0:
+		data["secret_id"] = string(rawSecretID)
```

Comment:
We trim off spaces in the check above.. Should we be trimming the string value that we set in the `data` map; i.e., `data["secret_id"] = strings.TrimSpace(string(rawSecretID))`?

Reply:
The rawSecretID is not trimmed anywhere, only read in from the file.
That trimming code is for cases where the file exists but only has a newline in it or something. We could trim the value set in data["secret_id"] as well. I had skipped it as it would be redundant.. but redundancy isn't necessarily bad. I'll add the extra string trim call.

Reply:
Added. Thanks!